torch_geometric/explain/algorithm/pg_explainer.py

```diff
@@ -0,0 +1,276 @@
+
+                    set_masks(self.model, self.edge_mask, edge_index_n)
+                    out = self.model(x=x_n, edge_index=edge_index_n,
+                                     **kwargs_n)
+                    loss += self.get_loss(out, target, mapping)
+                    clear_masks(self.model)
+
+                assert not torch.isnan(loss)
+                loss.backward()
+                optimizer.step()
+
+                if self.log:  # pragma: no cover
+                    pbar.update(1)
+
+        if self.log:
+            pbar.close()
+        clear_masks(self.model)
+        model.train(model_state)
+        self.training_need = False
+
+    def _get_temp(self, e: int) -> float:
+        temp = self.coeffs['temp']
+        return temp[0] * ((temp[1] / temp[0])**(e / self.epochs))
+
+    def _create_explainer_input(self, edge_index, z, node_idx=None) -> Tensor:
+        rows, cols = edge_index
+        z_j, z_i = z[rows], z[cols]
+        task_level = self.model_config.task_level
+        if task_level == ModelTaskLevel.node:
+            z_node = z[node_idx].repeat(rows.size(0), 1)
+            return torch.cat([z_i, z_j, z_node], 1)
+        else:
+            return torch.cat([z_i, z_j], 1)
+
+    def _compute_edge_mask(self, edge_weight, temperature=1.0, bias=0.0,
+                           training=True):
+
+        if training:  # noise is added to edge_weight.
+            bias += 0.0001
+            eps = (bias -
+                   (1 - bias)) * torch.rand(edge_weight.size()) + (1 - bias)
+            eps = eps.to(edge_weight.device)
+            return (eps.log() -
+                    (1 - eps).log() + edge_weight).squeeze(dim=1) / temperature
+
+        else:
+            return edge_weight.squeeze(dim=1)
+
+    def _loss(self, log_logits, prediction, node_idx=None, batch=None,
+              edge_index=None):
+        if self.model_config.mode == ModelMode.regression:
+            if node_idx is not None:
+                loss = F.mse_loss(log_logits[node_idx], prediction[node_idx],
+                                  reduction='sum')
+            else:
+                loss = F.mse_loss(log_logits, prediction, reduction='sum')
+        else:
+            if node_idx is not None:
+                loss = F.nll_loss(log_logits[node_idx], prediction[node_idx])
+            else:
+                loss = F.nll_loss(log_logits, prediction, reduction='sum')
+        mask = self.edge_mask.sigmoid().squeeze()
+
+        # Regularization losses
+        size_loss = mask.sum() * self.coeffs['edge_size']
+        mask_ent_reg = -mask * mask.log() - (1 - mask) * (1 - mask).log()
+        mask_ent_loss = mask_ent_reg.mean() if batch is None else scatter_mean(
+            mask_ent_reg, batch[edge_index[0]]).sum()
+        mask_ent_loss *= self.coeffs['edge_ent']
+
+        return loss + size_loss + mask_ent_loss
+
+    @staticmethod
+    def _drop_isolated_nodes(node_idxs: Tensor, edge_index: Tensor) -> Tensor:
```

Comment:
This could be useful for other explainers, maybe we can move it to `utils` ?

Reply:
Good point.